core/src/main/kotlin/de/fraunhofer/iem/kpiCalculator/core/strategy/KpiCalculationStrategy.kt

```diff
@@ -98,9 +73,85 @@ internal abstract class BaseKpiCalculationStrategy : KpiCalculationStrategy {
             failed = failed,
             additionalWeight = additionalWeight,
-            hasIncompleteResults = incompleteResults.isNotEmpty() || failed.isNotEmpty()
+            hasIncompleteResults = hasIncompleteResults
         )
+    }
 
+    private data class SeparatedKpiResults(
+        val successScores: List<Pair<KpiCalculationResult.Success, Double>>,
+        val failedScores: List<Pair<KpiCalculationResult, Double>>,
+        val missingEdgeWeights: Double,
+        val hasIncompleteResults: Boolean
+    )
+
+    /**
+     * Sorts the given childScores into separate lists for Success and Failed results.
+     * The sorting of Incomplete results depends on the strict mode.
+     *
+     * In strict mode, we consider Incomplete results as failed, and they are not
+     * considered in the KPI calculation.
+     * In not strict mode, we cast Incomplete results to Success results and use
+     * them for further calculation.
+     *
+     * @param childScores the KPI results and their planned edge weights.
+     * @param strict mode which influences the separation.
+     *
+     * @return SeparatedKpiResults, which contain a List of success and failed scores,
+     * missingEdgeWeights (the sum of all failed node's edge weights),
+     * hasIncompleteResults (indicates whether the original childScores contained an
+     * Incomplete result)
+     */
+    private fun processChildScores(
+        childScores: Collection<Pair<KpiCalculationResult, Double>>,
+        strict: Boolean
+    ): SeparatedKpiResults {
+
+        val successScores = mutableListOf<Pair<KpiCalculationResult.Success, Double>>()
+        val failedScores = mutableListOf<Pair<KpiCalculationResult, Double>>()
+        var missingEdgeWeight = 0.0
+        var hasIncompleteResults = false
+
+        childScores.forEach { childScore ->
+
+            // we need to extract the pair into a single variable to enable
+            // smart type casting in the following
+            when (val calcResult = childScore.first) {
+                is KpiCalculationResult.Success -> {
+                    // type erasure can't handle parameterized types so smart
+                    // casting won't work
+                    @Suppress("UNCHECKED_CAST")
+                    successScores.add(childScore as Pair<KpiCalculationResult.Success, Double>)
+                }
+
+                is KpiCalculationResult.Incomplete -> {
+                    if (strict) {
+                        failedScores.add(childScore)
+                        missingEdgeWeight += childScore.second
+                    } else {
+                        successScores.add(
+                            Pair(
+                                KpiCalculationResult.Success(score = calcResult.score),
+                                childScore.second
+                            )
+                        )
+                    }
+                    hasIncompleteResults = true
```

Comment:
instead of setting this explicitly you could make it an auto-computed property on the `SeparatedKpiResults` and check if either list does not has elements. (or if `missingEdgeWeights` is > 0.0 [if that's a valid constraint])

Reply:
actually no. 
Neither the `failed` list nor `missingEdgeWeights` tells us this information. 
This flag should indicate if there was any incomplete (Incomplete, Failed, Error) element in the initial childScores. However, in non strict mode, we cast Incomplete elements to Success elements and continue calculation. Thus, the `failed` list can be empty in this case, the `missingEdgeWeights` is 0.0, however, there was an incomplete input. 
This happens for this input `childScores = listOf(KpiResult.Incomplete(..), 1.0)` `calculateKpi(childScores, strict = false)`. Here we want to continue our calculation due to `strict = false`, however, we still want to indicate that the calculation is based on an incomplete result. Thus the flag.


 This is something to maybe improve on later or somehow make it more explicit.